File: app/src/main/java/com/example/studioghibliapp/MainActivity.java
package com.example.studioghibliapp;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.Toast;

import com.example.studioghibliapp.models.Film;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements AdapterView.OnItemSelectedListener {
    LinearLayout layout;
    Toolbar toolbar;
    RecyclerView rvFilms;
    FilmAdapter adapter;
    Spinner itemsSelector;

    String[] itemsSelectorOptions = {"5", "10", "50"};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("Peliculas");
        
        setupAdapter();

        layout = findViewById(R.id.ll_layout);
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(rvFilms.getContext(), layout.getOrientation());
        rvFilms.addItemDecoration(dividerItemDecoration);

        //Getting the instance of Spinner and applying OnItemSelectedListener on it
        itemsSelector = findViewById(R.id.sp_items_selector);
        itemsSelector.setOnItemSelectedListener((AdapterView.OnItemSelectedListener) this);

        //Creating the ArrayAdapter instance having the country list
        ArrayAdapter aa = new ArrayAdapter(this, android.R.layout.simple_spinner_item, itemsSelectorOptions);
        aa.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        //Setting the ArrayAdapter data on the Spinner
        itemsSelector.setAdapter(aa);

    }

    private void setupAdapter() {
        rvFilms = findViewById(R.id.rv_films);

        adapter = new FilmAdapter(getFilms(), new FilmAdapter.OnItemClickListener() {
            @Override
            public void onItemClickListener(Film film) {
                // Funcionalidad
                // Toast.makeText(MainActivity.this, film.getTitle(), Toast.LENGTH_LONG).show();
                Intent filmDetailsActivity = new Intent(MainActivity.this, FilmDetailsActivity.class);
                startActivity(filmDetailsActivity);
            }
        });
        rvFilms.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu){
        getMenuInflater().inflate(R.menu.menu_main,menu);
        return super.onCreateOptionsMenu(menu);
    }
    @Override
    public boolean onOptionsItemSelected (MenuItem item){
        switch (item.getItemId()){
            //Realizar accion
            case R.id.item_exit:
                Toast.makeText(this, "Sesión cerrada", Toast.LENGTH_SHORT).show();
                Intent intent = new Intent(MainActivity.this, LoginActivity.class);
                startActivity(intent);
                break;

            case R.id.item_Personajes:
                Intent intent0 = new Intent (MainActivity.this, PersonajesActivity.class );
                startActivity(intent0);
                break;

            case R.id.item_Vehiculos:
                Intent intent1 = new Intent (MainActivity.this, VehiculosActivity.class );
                startActivity(intent1);
                break;

            case R.id.item_Especies:
                Intent intent2 = new Intent (MainActivity.this, EspeciesActivity.class );
                startActivity(intent2);
                break;

            case R.id.item_Lugares:
                Intent intent3 = new Intent (MainActivity.this, LugaresActivity.class );
                startActivity(intent3);
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    private List<Film> getFilms() {
        return new ArrayList<Film>() {{
            add(new Film("1", "Castle in the Sky","https://image.tmdb.org/t/p/w600_and_h900_bestv2/npOnzAbLh6VOIu3naU5QaEcTepo.jpg" ,"The orphan Sheeta inherited a mysterious crystal that links her to the mythical sky-kingdom of Laputa. With the help of resourceful Pazu and a rollicking band of sky pirates, she makes her way to the ruins of the once-great civilization. Sheeta and Pazu must outwit the evil Muska, who plans to use Laputa's science to make himself ruler of the world.", "Hayao Miyazaki", "1986"));
            add(new Film("2", "Grave of the Fireflies", "https://image.tmdb.org/t/p/w600_and_h900_bestv2/qG3RYlIVpTYclR9TYIsy8p7m7AT.jpg", "In the latter part of World War II, a boy and his sister, orphaned when their mother is killed in the firebombing of Tokyo, are left to survive on their own in what remains of civilian life in Japan. The plot follows this boy and his sister as they do their best to survive in the Japanese countryside, battling hunger, prejudice, and pride in their own quiet, personal battle.", "Isao Takahata", "1988"));
            add(new Film("3", "Castle in the Sky","https://image.tmdb.org/t/p/w600_and_h900_bestv2/npOnzAbLh6VOIu3naU5QaEcTepo.jpg" ,"The orphan Sheeta inherited a mysterious crystal that links her to the mythical sky-kingdom of Laputa. With the help of resourceful Pazu and a rollicking band of sky pirates, she makes her way to the ruins of the once-great civilization. Sheeta and Pazu must outwit the evil Muska, who plans to use Laputa's science to make himself ruler of the world.", "Hayao Miyazaki", "1986"));
            add(new Film("4", "Grave of the Fireflies", "https://image.tmdb.org/t/p/w600_and_h900_bestv2/qG3RYlIVpTYclR9TYIsy8p7m7AT.jpg", "In the latter part of World War II, a boy and his sister, orphaned when their mother is killed in the firebombing of Tokyo, are left to survive on their own in what remains of civilian life in Japan. The plot follows this boy and his sister as they do their best to survive in the Japanese countryside, battling hunger, prejudice, and pride in their own quiet, personal battle.", "Isao Takahata", "1988"));
            add(new Film("5", "Castle in the Sky","https://image.tmdb.org/t/p/w600_and_h900_bestv2/npOnzAbLh6VOIu3naU5QaEcTepo.jpg" ,"The orphan Sheeta inherited a mysterious crystal that links her to the mythical sky-kingdom of Laputa. With the help of resourceful Pazu and a rollicking band of sky pirates, she makes her way to the ruins of the once-great civilization. Sheeta and Pazu must outwit the evil Muska, who plans to use Laputa's science to make himself ruler of the world.", "Hayao Miyazaki", "1986"));
            add(new Film("6", "Grave of the Fireflies", "https://image.tmdb.org/t/p/w600_and_h900_bestv2/qG3RYlIVpTYclR9TYIsy8p7m7AT.jpg", "In the latter part of World War II, a boy and his sister, orphaned when their mother is killed in the firebombing of Tokyo, are left to survive on their own in what remains of civilian life in Japan. The plot follows this boy and his sister as they do their best to survive in the Japanese countryside, battling hunger, prejudice, and pride in their own quiet, personal battle.", "Isao Takahata", "1988"));
            add(new Film("7", "Castle in the Sky","https://image.tmdb.org/t/p/w600_and_h900_bestv2/npOnzAbLh6VOIu3naU5QaEcTepo.jpg" ,"The orphan Sheeta inherited a mysterious crystal that links her to the mythical sky-kingdom of Laputa. With the help of resourceful Pazu and a rollicking band of sky pirates, she makes her way to the ruins of the once-great civilization. Sheeta and Pazu must outwit the evil Muska, who plans to use Laputa's science to make himself ruler of the world.", "Hayao Miyazaki", "1986"));
            add(new Film("8", "Grave of the Fireflies", "https://image.tmdb.org/t/p/w600_and_h900_bestv2/qG3RYlIVpTYclR9TYIsy8p7m7AT.jpg", "In the latter part of World War II, a boy and his sister, orphaned when their mother is killed in the firebombing of Tokyo, are left to survive on their own in what remains of civilian life in Japan. The plot follows this boy and his sister as they do their best to survive in the Japanese countryside, battling hunger, prejudice, and pride in their own quiet, personal battle.", "Isao Takahata", "1988"));
            add(new Film("9", "Castle in the Sky","https://image.tmdb.org/t/p/w600_and_h900_bestv2/npOnzAbLh6VOIu3naU5QaEcTepo.jpg" ,"The orphan Sheeta inherited a mysterious crystal that links her to the mythical sky-kingdom of Laputa. With the help of resourceful Pazu and a rollicking band of sky pirates, she makes her way to the ruins of the once-great civilization. Sheeta and Pazu must outwit the evil Muska, who plans to use Laputa's science to make himself ruler of the world.", "Hayao Miyazaki", "1986"));
            add(new Film("10", "Grave of the Fireflies", "https://image.tmdb.org/t/p/w600_and_h900_bestv2/qG3RYlIVpTYclR9TYIsy8p7m7AT.jpg", "In the latter part of World War II, a boy and his sister, orphaned when their mother is killed in the firebombing of Tokyo, are left to survive on their own in what remains of civilian life in Japan. The plot follows this boy and his sister as they do their best to survive in the Japanese countryside, battling hunger, prejudice, and pride in their own quiet, personal battle.", "Isao Takahata", "1988"));
            add(new Film("11", "Grave of the Fireflies", "https://image.tmdb.org/t/p/w600_and_h900_bestv2/qG3RYlIVpTYclR9TYIsy8p7m7AT.jpg", "In the latter part of World War II, a boy and his sister, orphaned when their mother is killed in the firebombing of Tokyo, are left to survive on their own in what remains of civilian life in Japan. The plot follows this boy and his sister as they do their best to survive in the Japanese countryside, battling hunger, prejudice, and pride in their own quiet, personal battle.", "Isao Takahata", "1988"));
            add(new Film("12", "Castle in the Sky","https://image.tmdb.org/t/p/w600_and_h900_bestv2/npOnzAbLh6VOIu3naU5QaEcTepo.jpg" ,"The orphan Sheeta inherited a mysterious crystal that links her to the mythical sky-kingdom of Laputa. With the help of resourceful Pazu and a rollicking band of sky pirates, she makes her way to the ruins of the once-great civilization. Sheeta and Pazu must outwit the evil Muska, who plans to use Laputa's science to make himself ruler of the world.", "Hayao Miyazaki", "1986"));
        }};
    }

    //Performing action onItemSelected and onNothing selected
    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int position, long l) {
        //Toast.makeText(getApplicationContext(),itemsSelectorOptions[position], Toast.LENGTH_LONG).show();
    }

    @Override
    public void onNothingSelected(AdapterView<?> adapterView) {
        // TODO Auto-generated method stub
    }
}

File: app/src/main/java/com/example/studioghibliapp/MainActivity.kt
package com.example.studioghibliapp

import com.example.studioghibliapp.RetrofitClient.retrofit
import androidx.appcompat.app.AppCompatActivity
import android.widget.AdapterView
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import android.widget.Spinner
import android.os.Bundle
import android.widget.ArrayAdapter
import android.content.Intent
import androidx.recyclerview.widget.DividerItemDecoration
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.lifecycleScope
import com.example.studioghibliapp.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity(), AdapterView.OnItemSelectedListener {
    lateinit var layout: LinearLayout
    lateinit var toolbar: Toolbar
    lateinit var rvItems: RecyclerView
    private lateinit var itemsSelector: Spinner

    var itemsSelectorOptions = arrayOf<Number?>(5, 10, 50)

    private val api: StudioGhibliAPI = retrofit.create(StudioGhibliAPI::class.java)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.title = "Peliculas"

        rvItems = findViewById(R.id.rv_items)

        setupAdapter()
        setupAdapterFilms()

        //Getting the instance of Spinner and applying OnItemSelectedListener on it
        itemsSelector = findViewById(R.id.sp_items_selector)
        itemsSelector.onItemSelectedListener = this

        //Creating the ArrayAdapter instance having the country list
        val aa: ArrayAdapter<*> = ArrayAdapter<Any?>(this, android.R.layout.simple_spinner_item, itemsSelectorOptions)
        aa.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)

        //Setting the ArrayAdapter data on the Spinner
        itemsSelector.adapter = aa

        itemsSelector.setSelection(2); // 50
    }

    private fun setupAdapterFilms() {
        var responseList: MutableList<Film> = mutableListOf()

        val callGetFilms = api.getFilms()
        callGetFilms.enqueue(object: Callback<List<Film>?> {
            override fun onResponse(call: Call<List<Film>?>, response: Response<List<Film>?>) {
                val filmsRest = response.body()

                var counter = 0;
                filmsRest?.forEach {
                    if (counter < itemsSelector.selectedItem.toString().toInt()){
                        val film = Film(
                            it.id,
                            it.title,
                            it.image,
                            it.movie_banner,
                            it.description,
                            it.director,
                            it.release_date,
                            it.producer,
                            it.rt_score
                        )
                        responseList.add(film)
                        counter++
                    }
                }

                FilmAdapter(responseList) {
                    val filmDetailsActivity = Intent(this@MainActivity, FilmDetailsActivity::class.java)
                    filmDetailsActivity.putExtra("FilmID", it.id)
                    startActivity(filmDetailsActivity)
                }.let {
                    rvItems.adapter = it
                }
            }

            override fun onFailure(call: Call<List<Film>?>, t: Throwable) {
                Log.e("REST", t.message?: "")
            }
        })
    }

    private fun setupAdapterPeopleList() {
        var responseList: MutableList<People> = mutableListOf()

        val callGetPeopleList = api.getPeople()
        callGetPeopleList.enqueue(object: Callback<List<People>?> {
            override fun onResponse(call: Call<List<People>?>, response: Response<List<People>?>) {
                val peopleRest = response.body()

                var counter = 0;
                peopleRest?.forEach {
                    if (counter < itemsSelector.selectedItem.toString().toInt()) {
                        var people = People(it.id, it.name, it.gender, it.age)
                        responseList.add(people)
                        counter++
                    }
                }

                PeopleAdapter(responseList) {}.let {
                    rvItems.adapter = it
                }
            }

            override fun onFailure(call: Call<List<People>?>, t: Throwable) {
                Log.e("REST", t.message?: "")
            }
        })
    }

    private fun setupAdapterLocations() {
        var responseList: MutableList<Location> = mutableListOf()

        val callGetLocations = api.getLocations()
        callGetLocations.enqueue(object: Callback<List<Location>?> {
            override fun onResponse(call: Call<List<Location>?>, response: Response<List<Location>?>) {
                val locationsRest = response.body()

                var counter = 0
                locationsRest?.forEach {
                    if (counter < itemsSelector.selectedItem.toString().toInt()) {
                        var location = Location(it.id, it.name)
                        responseList.add(location)
                        counter++
                    }
                }

                LocationAdapter(responseList) {}.let {
                    rvItems.adapter = it
                }
            }

            override fun onFailure(call: Call<List<Location>?>, t: Throwable) {
                Log.e("REST", t.message?: "")
            }
        })
    }

    private fun setupAdapterVehicles() {
        var responseList: MutableList<Vehicle> = mutableListOf()

        val callGetVehicles = api.getVehicles()
        callGetVehicles.enqueue(object: Callback<List<Vehicle>?> {
            override fun onResponse(call: Call<List<Vehicle>?>, response: Response<List<Vehicle>?>) {
                val vehiclesRest = response.body()

                var counter = 0
                vehiclesRest?.forEach {
                    if (counter < itemsSelector.selectedItem.toString().toInt()) {
                        var vehicle = Vehicle(it.id, it.name, it.description, it.vehicle_class)
                        responseList.add(vehicle)
                        counter++
                    }
                }

                VehicleAdapter(responseList) {}.let {
                    rvItems.adapter = it
                }
            }

            override fun onFailure(call: Call<List<Vehicle>?>, t: Throwable) {
                Log.e("REST", t.message?: "")
            }
        })
    }

    private fun setupAdapterSpeciesList() {
        var responseList: MutableList<Species> = mutableListOf()

        val callGetSpeciesList = api.getSpecies()
        callGetSpeciesList.enqueue(object: Callback<List<Species>?> {
            override fun onResponse(call: Call<List<Species>?>, response: Response<List<Species>?>) {
                val speciesListRest = response.body()

                var counter = 0
                speciesListRest?.forEach {
                    if (counter < itemsSelector.selectedItem.toString().toInt()) {
                        var species = Species(it.id, it.name, it.classification)
                        responseList.add(species)
                        counter++
                    }
                }

                SpeciesAdapter(responseList) {}.let {
                    rvItems.adapter = it
                }
            }

            override fun onFailure(call: Call<List<Species>?>, t: Throwable) {
                Log.e("REST", t.message?: "")
            }
        })
    }

    private fun setupAdapter() {
        var emptyFilmList: MutableList<Film> = mutableListOf()

        lifecycleScope.launch {
            withContext(Dispatchers.IO) {
                FilmAdapter(emptyFilmList) {}.let {
                    rvItems.adapter = it
                }
            }
        }

        layout = findViewById(R.id.ll_layout)
        val dividerItemDecoration = DividerItemDecoration(rvItems.context, layout.orientation)
        rvItems.addItemDecoration(dividerItemDecoration)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.item_exit -> {
                val sp = applicationContext.getSharedPreferences(Consts.SP_CREDENTIALS, MODE_PRIVATE)
                val spEditor = sp.edit()
                spEditor.putString(Consts.USER, null)
                spEditor.putString(Consts.PASSWORD, null)
                spEditor.apply()
                Toast.makeText(this, "Sesión cerrada", Toast.LENGTH_SHORT).show()
                val intent = Intent(this@MainActivity, LoginActivity::class.java)
                startActivity(intent)
            }
            R.id.item_films -> {
                setupAdapterFilms();
                supportActionBar!!.title = "Peliculas"
            }
            R.id.item_people -> {
                setupAdapterPeopleList();
                supportActionBar!!.title = "Personajes"
            }
            R.id.item_vehicles -> {
                setupAdapterVehicles()
                supportActionBar!!.title = "Vehículos"
            }
            R.id.item_species -> {
                setupAdapterSpeciesList()
                supportActionBar!!.title = "Especies"
            }
            R.id.item_locations -> {
                setupAdapterLocations();
                supportActionBar!!.title = "Lugares"
            }
        }
        return super.onOptionsItemSelected(item)
    }

    //Performing action onItemSelected and onNothing selected
    override fun onItemSelected(adapterView: AdapterView<*>?, view: View, position: Int, l: Long) {
        if(supportActionBar!!.title === "Peliculas") {
            setupAdapterFilms()
        } else if(supportActionBar!!.title === "Personajes") {
            setupAdapterPeopleList();
        } else if(supportActionBar!!.title === "Vehículos") {
            setupAdapterVehicles()
        } else if(supportActionBar!!.title === "Especies") {
            setupAdapterSpeciesList()
        } else if(supportActionBar!!.title === "Lugares") {
            setupAdapterLocations();
        }
    }

    override fun onNothingSelected(adapterView: AdapterView<*>?) {
        // TODO Auto-generated method stub
    }
}
